Public/Src/Engine/Dll/EngineSchedule.cs

```diff
@@ -806,19 +810,95 @@ private bool ShouldSerializeOptimizationDataStructurePostExecution()
                     // Everything that is not an output under a shared opaque is considered part of the build. 
                     isPathInBuild: path =>
                         // Scheduler.PipGraph.IsPathInBuild is used for extra safety.
-                        Scheduler.PipGraph.IsPathInBuild(AbsolutePath.Create(Context.PathTable, path)) ||
+                        scheduler.PipGraph.IsPathInBuild(AbsolutePath.Create(scheduler.Context.PathTable, path)) ||
                         !SharedOpaqueOutputHelper.IsSharedOpaqueOutput(path), 
-                    pathsToScrub: sharedOpaqueDirectories.Select(directory => directory.Path.ToString(Context.PathTable)),
+                    pathsToScrub: sharedOpaqueDirectories.Select(directory => directory.Path.ToString(scheduler.Context.PathTable)),
                     blockedPaths: nonScrubbablePaths,
                     nonDeletableRootDirectories: outputDirectories,
                     // Mounts don't need to be scrubbable for this operation to take place.
                     mountPathExpander: null,
                     maxDegreeParallelism: Environment.ProcessorCount,
-                    tempDirectoryCleaner: m_tempCleaner);
+                    tempDirectoryCleaner: tempCleaner);
 
                 Logger.Log.ScrubbingSharedOpaquesStarted(loggingContext);
-                scrubber.RemoveExtraneousFilesAndDirectories(Context.CancellationToken);
+                scrubber.RemoveExtraneousFilesAndDirectories(scheduler.Context.CancellationToken);
+            }
+        }
+
+        internal static IReadOnlyList<string> GetNonScrubbablePaths(

```

Comment:
same as before, just moved from `Engine` to `EngineSchedule` so that it can be called from integration tests